### PR TITLE
 - Refactored Wallet and Inventory code

### PR DESCRIPTION
**Pull request by sebastian24**

### CODE REVIEW

File: spilsdk/src/main/java/com/spilgames/spilsdk/events/response/PlayerDataResponseEvent.java [hunk 1]
```diff
@@ -31,7 +30,7 @@
                     if(walletJSON.has("currencies")){
                         JSONArray currenciesJSON = walletJSON.getJSONArray("currencies");
 
-                        ArrayList<PlayerCurrency> currencyArrayList = new ArrayList<>();
+                        HashMap<Integer, PlayerCurrency> currencyArrayList = new HashMap<>();
```
Comment (sebastian24) [inline spilsdk/src/main/java/com/spilgames/spilsdk/events/response/PlayerDataResponseEvent.java]: Yes you are right. It does make more sense. It was not changed when I updated. I will do that now

File: spilsdk/src/main/java/com/spilgames/spilsdk/events/response/PlayerDataResponseEvent.java [hunk 2]
```diff
@@ -60,7 +59,7 @@
                     if(inventoryJSON.has("items")){
                         JSONArray itemsJSON = inventoryJSON.getJSONArray("items");
 
-                        ArrayList<PlayerItem> itemsArrayList = new ArrayList<>();
+                        HashMap<Integer, PlayerItem> itemsArrayList = new HashMap<>();
```
Comment (sebastian24) [inline spilsdk/src/main/java/com/spilgames/spilsdk/events/response/PlayerDataResponseEvent.java]: Same as the other one

File: spilsdk/src/main/java/com/spilgames/spilsdk/gamedata/SpilGameDataUtil.java [hunk 3]
```diff
@@ -204,73 +178,51 @@ public String getShop(){
...
-            if(valid){
-                return gson.toJson(spilGameData.getShop());
-            } else {
-                SpilSdk.getInstance(context).getGameDataCallbacks().gameDataError(ErrorCodes.LoadFailed);
-                requestGameData();
-                return null;
-            }
+            return gson.toJson(spilGameData.getShop());
 
         } else {
```
Comment (sebastian24) [inline spilsdk/src/main/java/com/spilgames/spilsdk/gamedata/SpilGameDataUtil.java]: will flip it. again overlooked :)

File: spilsdk/src/main/java/com/spilgames/spilsdk/gamedata/SpilGameDataUtil.java [hunk 4]
```diff
@@ -204,73 +178,51 @@ public String getShop(){
...
-            if(valid){
-                return gson.toJson(spilGameData.getPromotions());
-            } else {
-                SpilSdk.getInstance(context).getGameDataCallbacks().gameDataError(ErrorCodes.LoadFailed);
-                requestGameData();
-                return null;
-            }
+            return gson.toJson(spilGameData.getPromotions());
 
         } else {
```
Comment (sebastian24) [inline spilsdk/src/main/java/com/spilgames/spilsdk/gamedata/SpilGameDataUtil.java]: same

File: spilsdk/src/main/java/com/spilgames/spilsdk/events/response/PlayerDataResponseEvent.java [hunk 5]
```diff
@@ -71,9 +70,9 @@
 
                             item.setDelta(jsonObject.getInt("delta"));
 
-                            itemsArrayList.add(item);
+                            itemsHashMap.put(item.getId(), item);
                         }
-                        inventoryObject.setItems(itemsArrayList);
+                        inventoryObject.setItemsMap(itemsHashMap);
```
Comment (svenvanhoenderdal) [inline spilsdk/src/main/java/com/spilgames/spilsdk/events/response/PlayerDataResponseEvent.java]: preferably: setItems(

File: spilsdk/src/main/java/com/spilgames/spilsdk/events/response/PlayerDataResponseEvent.java [hunk 6]
```diff
@@ -104,10 +103,10 @@
 
                             currency.setDelta(jsonObject.getInt("delta"));
 
-                            currencyArrayList.add(currency);
+                            currencyHashMap.put(currency.getId(), currency);
                         }
 
-                        walletObject.setCurrencies(currencyArrayList);
+                        walletObject.setCurrenciesMap(currencyHashMap);
```
Comment (svenvanhoenderdal) [inline spilsdk/src/main/java/com/spilgames/spilsdk/events/response/PlayerDataResponseEvent.java]: preferably: setCurrencies(

File: spilsdk/src/main/java/com/spilgames/spilsdk/playerdata/PlayerDataUtil.java [hunk 7]
```diff
@@ -101,155 +104,152 @@ public void processPlayerData(Wallet wallet, Inventory inventory){
...
-                    userProfile.getWallet().getCurrencies().get(i).setDelta(0);
-                }
+        if(userProfile == null){
+            return;
+        }
 
-                //Check if the received offset is higher than the current offset and if the sent currencies list is not empty
-                if(userProfile.getWallet().getOffset() < wallet.getOffset() && !wallet.getCurrencies().isEmpty()) {
+        //Checks if the wallet is empty and if so throws an error to the developer
+        if(receivedWallet != null){
```
Comment (svenvanhoenderdal) [inline spilsdk/src/main/java/com/spilgames/spilsdk/playerdata/PlayerDataUtil.java]: If you extract these parts to methods you can flatten it (and make it more readable).

**Pull request merged**